plugins/qsi-dock/fu-qsi-dock-mcu-device.c

```diff
@@ -0,0 +1,703 @@
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_qsi_dock_mcu_device_checksum(FuQsiDockMcuDevice *self, guint32 checksum, guint32 check_length, GError **error)
+{
+	guint8 buf[64] = {0};
+	
+	buf[0] = Report_ID;
+	buf[1] = CmdPrimary_CMD_SPI;
+	buf[2] = CmdSecond_SPI_External_Flash_Checksum;
+
+        buf[3] = (check_length) & 0xff;
+        buf[4] = (check_length >> 8) & 0xff;
+        buf[5] = (check_length >> 16) & 0xff;
+        buf[6] = (check_length >> 24) & 0xff;
+        
+	buf[7] = (checksum) & 0xff;
+        buf[8] = (checksum >> 8) & 0xff;
+
+	// Set Report	
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+					Report_ID,
+					buf,
+					sizeof(buf),
+					FU_QSI_DOCK_MCU_DEVICE_TIMEOUT,
+					FU_HID_DEVICE_FLAG_NONE,
+					error))
+			return FALSE;
+
+	// GetReport 
+	memset(buf, 0x0, sizeof(buf));
+	if (!fu_hid_device_get_report(FU_HID_DEVICE(self),
+					Report_ID,
+					buf,
+					sizeof(buf),
+					FU_QSI_DOCK_MCU_DEVICE_TIMEOUT,
+					FU_HID_DEVICE_FLAG_NONE,
+					error)) 
+			return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_qsi_dock_mcu_device_write_chunk(FuQsiDockMcuDevice *self, FuChunk *chk, GError **error, guint32 *checksum_tmp)
+{
+	guint8 buf[64] = {0};
+	guint32 length = 60;
+	guint32 pagesize = fu_chunk_get_data_sz(chk);
+	guint16 page_data_count = 256;
+
+	while (pagesize != 0 || page_data_count != 0) {
+		memset(buf, 0x0, sizeof(buf));
+		buf[0] = Report_ID;
+		buf[1] = CmdPrimary_Mass_SPI;
+
+		/* set length and buffer */
+		if (pagesize > TX_ISP_LENGTH_MCU) {
+			length = TX_ISP_LENGTH_MCU;
+			pagesize -= TX_ISP_LENGTH_MCU;
+			page_data_count -= TX_ISP_LENGTH_MCU;
+			buf[2] = length;
+		}else{
+			length = pagesize;
+			if(page_data_count == pagesize){
+				page_data_count = 0;
+				buf[2] = length;
+			}else if(page_data_count >= TX_ISP_LENGTH_MCU){
+				page_data_count -= TX_ISP_LENGTH_MCU;
+				buf[2] =  TX_ISP_LENGTH_MCU;
+			}else{
+				buf[2] = page_data_count;
+				page_data_count = 0;
+			}
+			pagesize = 0;
+		}
+
+		// SetReport
+		if (!fu_memcpy_safe(buf,
+				    sizeof(buf),
+				    0x04, /* dst */
+				    fu_chunk_get_data(chk),
+				    fu_chunk_get_data_sz(chk),
+				    fu_chunk_get_data_sz(chk) - pagesize - length, /* src */
+				    length,
+				    error))
+			return FALSE;
+		
+		// Checksum value
+		for(int i=4;i<64;i++){
+			*checksum_tmp +=buf[i];
+		}
+		
+		if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+					      Report_ID,
+					      buf,
+					      sizeof(buf),
+					      FU_QSI_DOCK_MCU_DEVICE_TIMEOUT,
+					      FU_HID_DEVICE_FLAG_NONE,
+					      error))
+			return FALSE;
+
+		// GetReport 
+		memset(buf, 0x0, sizeof(buf));
+		if (!fu_hid_device_get_report(FU_HID_DEVICE(self),
+					      Report_ID,
+					      buf,
+					      sizeof(buf),
+					      FU_QSI_DOCK_MCU_DEVICE_TIMEOUT,
+					      FU_HID_DEVICE_FLAG_NONE,
+					      error)) 
+			return FALSE;
+
+		if(!buf[2] == 0) {
+			return FALSE;
```

Comment:
if you return FALSE you need to use g_set_error() to set an error, otherwise the daemon explodes.